input-group: add validation section (#3444)
IgniteUI/igniteui-docfx: en/components/input-group.md

           iframe-src="{environment:demosBaseUrl}/data-entries/typed-form" >
</code-view>

## Validation

> The following samples demonstrate how input validation could be achieved when using template-driven or reactive forms.

Reviewer: ```suggestion
The following samples demonstrate how to configure input validation when using [template-driven](https://angular.io/guide/forms) or [reactive forms](https://angular.io/guide/reactive-forms).
```

### Template-Driven Forms
Template-driven form validation is achieved by adding validation attributes, i.e., `required`, `minLength`, etc., to the `input` element.

```html

> <igx-input-group>

Reviewer: Include the wrapping form element in each code snippet, together with the submit button.

    <label igxLabel for="username">Username</label>
    <input igxInput name="username" type="text" required />
</igx-input-group>

<igx-input-group>
    <label igxLabel for="email">Email</label>
    <input igxInput name="email" type="email" required email />
</igx-input-group>

<igx-input-group>
    <label igxLabel for="password">Password</label>
    <input igxInput name="password" type="password" required minlength="8" />
</igx-input-group>
```

The `required` attribute adds an asterisk next to the label, indicating that this field must be completed, however, when having additional validation like `email` and `minLength`, the user could be notified via the [`igx-hint`]({environment:angularApiUrl}/classes/igxhintdirective.html) directive.

> The following example uses two-way data binding and demonstrates how to inspect the control's state by exporting the `ngModel` to a local variable.

Reviewer: Also include an example on how to inspect general form validity through the ngForm. In general the check there would be perform in the submit button to make it enabled/disabled.

</igx-input-group>
```

The result from the above configurations could be seen in the below sample. Start typing into the Email and Password fields and you will notice that the `<igx-hint>` is shown if the entered values are invalid.

<code-view style="height:480px"
           data-demos-base-url="{environment:demosBaseUrl}"
           iframe-src="{environment:demosBaseUrl}/data-entries/template-driven-form-validation" >
</code-view>

### Reactive Forms
Reactive form validation is achieved by adding validator functions directly to the form control model in the component class. After creating the control in the component class, it should be associated with a form control element in the template.

```ts
public registrationForm = new FormGroup<User>({
    username: new FormControl<string>('', { nonNullable: true, validators: [Validators.required] }),
    email: new FormControl<string>('', { nonNullable: true, validators: [Validators.required, Validators.email] }),
    password: new FormControl<string>('', { nonNullable: true, validators: [Validators.required, Validators.minLength(8)] })
});
```

```html

> <form [formGroup]="registrationForm">

Reviewer: Again, include a submit button.

The following example demonstrates how to access the control through a `get` method and inspect its state.

```ts
get email() {
    return this.registrationForm.get('email');
}
get password() {
    return this.registrationForm.get('password');
}
```

```html
<igx-input-group>
    <label igxLabel for="email">Email</label>
    <input igxInput name="email" type="email" formControlName="email" />
    <igx-hint *ngIf="email.errors?.email">Please enter a valid email</igx-hint>
</igx-input-group>

<igx-input-group>
    <label igxLabel for="password">Password</label>
    <input igxInput name="password" type="password" formControlName="password" />
    <igx-hint *ngIf="password.errors?.minlength">Password should be at least 8 characters</igx-hint>
</igx-input-group>
```

The result from the above configurations could be seen in the below sample.

<code-view style="height:480px"
           data-demos-base-url="{environment:demosBaseUrl}"
           iframe-src="{environment:demosBaseUrl}/data-entries/reactive-form-validation" >
</code-view>

### Custom Validators

> Some input fields may require custom validation and this could be achieved via custom validators. When the value is valid, the validator will return `null` and when the value is invalid, it will generate a set of errors that could be used to display a specific error message on the screen.

Reviewer: I have left a review comment about the custom validation sample in the samples repo. There, you would be able to create an example of how one from control drives the validity of a different form control, or of the entire group.
